accepted/0000-improve-command-suggestions.md

```diff
@@ -0,0 +1,104 @@
+
+A critical part of this puzzle to avoid breaking things for folks is that we default to not assuming we know what the user wants. The affordances in the npm CLI (like `npm hlep` working as `npm help`) are, in my opinion, a mistake. When a user makes a typo, we should not assume that we know what they're trying to do. We should tell them their command was invalid and suggest some valid commands that may be what they wanted.
+
+## Detailed Explanation
+
+As a starting point, I'd like to see this apply to the output when an invalid command is encountered as well as when no script matches in `npm run` and no binary matches when running `npm exec`. Any of these three scenarios should suggest valid `npm run` or `npm exec` commands if one seems available.
+
+Current behavior of `npm run tes` when a script named `"tes"` is not defined, but one named `"test"` is:
+
+```
+npm ERR! missing script: tes
+npm ERR!
+npm ERR! Did you mean this?
+npm ERR!     test
+```
+
+Proposed output with the same scenario, but the addition of a package in the tree named `test` that includes a discoverable binary:
+```
+npm ERR! missing script: tes
+npm ERR!
+npm ERR! Did you mean one of these?
+npm ERR!     npm run test
+npm ERR!     npm exec -- test
+```
+
+Current behavior of `npm foo` where `foo` is an unknown command:
+
+```
+Top hits for "foo"
+————————————————————————————————————————————————————————————————————————————————
+npm help package-json                                                     foo:50
+npm help exec                                                             foo:16
+npm help npx                                                              foo:16
+npm help folders                                                          foo:15
+npm help scripts                                                           foo:7
+npm help config                                                            foo:7
+npm help init                                                              foo:6
+npm help developers                                                        foo:4
+npm help star                                                             foo:3
+npm help package-lock-json                                                 foo:3
+————————————————————————————————————————————————————————————————————————————————
+(run with -l or --long to see more context)t
+```
+
+Proposed behavior of `npm foo` when both a script named `"foo"` and a package named `foo` with a discoverable bin exist:
+```
+npm ERR! unknown command: foo
+npm ERR!
+npm ERR! Did you mean one of these?
+npm ERR!     npm run foo
+npm ERR!     npm exec -- foo
```

Comment:
```suggestion
npm ERR! Did you mean one of these?
npm ERR!     npm run foo        # (runs the "foo" script)
npm ERR!     npm exec -- foo    # (runs the "foo" binary in node_modules)
```
maybe something like this?

Reply:
great idea, each suggested command can come with a description